### ISSUE TITLE
"R CMD check" failure on Linux ARM64

### ISSUE THREAD

**Issue by martin-g** [functionalgenomics/atena#6]
Hello,

There is an initiative to support Linux ARM64 for Bioconductor.
While testing the packages we faced this failure:

```
biocbuild@kunpeng1 ~/git> R CMD check atena_1.5.10.tar.gz 
* using log directory ‘/home/biocbuild/git/atena.Rcheck’
* using R Under development (unstable) (2023-03-12 r83975)
* using platform: aarch64-unknown-linux-gnu (64-bit)
* R was compiled by
    gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
    GNU Fortran (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
* running under: Ubuntu 22.04.2 LTS
* using session charset: UTF-8
* checking for file ‘atena/DESCRIPTION’ ... OK
* checking extension type ... Package
* this is package ‘atena’ version ‘1.5.10’
* package encoding: UTF-8
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking whether package ‘atena’ can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking ‘build’ directory ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking startup messages can be suppressed ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... OK
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking installed files from ‘inst/doc’ ... OK
* checking files in ‘vignettes’ ... OK
* checking examples ... ERROR
Running examples in ‘atena-Ex.R’ failed
The error most likely occurred in:

> ### Name: annotaTEs
> ### Title: Get RepeatMasker UCSC annotations
> ### Aliases: annotaTEs
> 
> ### ** Examples
> 
> rmsk_gr <- annotaTEs(genome = "hg19", parsefun = rmskidentity)
snapshotDate(): 2023-03-21
loading from cache
Error: failed to load resource
  name: AH99002
  title: UCSC RepeatMasker annotations (Mar2020) for Human (hg19)
  reason: error reading from connection
Execution halted
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ...
  Running ‘runTests.R’
 ERROR
Running the tests in ‘tests/runTests.R’ failed.
Last 13 lines of output:
  1 Test Suite : 
  atena RUnit Tests - 13 test functions, 0 errors, 1 failure
  FAILURE in test_TEtranscripts: Error in checkEqualsNumeric(head(sort(assay(ttSE), decreasing = TRUE)),  : 
    Mean relative difference: 0.007352941
  
  
  Test files with failing tests
  
     test_TEtranscripts.R 
       test_TEtranscripts 
  
  
  Error in BiocGenerics:::testPackage("atena") : 
    unit tests failed for package atena
  Execution halted
* checking for unstated dependencies in vignettes ... OK
* checking package vignettes in ‘inst/doc’ ... OK
* checking running R code from vignettes ...
  ‘atena.Rmd’ using ‘UTF-8’... OK
 NONE
* checking re-building of vignette outputs ... OK
* checking PDF version of manual ... OK
* DONE

Status: 2 ERRORs
See
  ‘/home/biocbuild/git/atena.Rcheck/00check.log’
for details.
```

Any idea why it fails ?
The check passes fine on x86_64 - https://bioconductor.org/checkResults/3.17/bioc-LATEST/atena but it fails consistently on aarch64.

**Comment by rcastelo** [comment]
Hi, thank you so much for your efforts to port Bioconductor packages to Linux ARM64. I see two different problems that may have different and unrelated causes:

1. The line:

```
rmsk_gr <- annotaTEs(genome = "hg19", parsefun = rmskidentity)
```

gives the error `failed to load resource` and this seems to be a problem with another package that atena uses, called `AnnotationHub`, could you try whether interactively this works:

```
library(AnnotationHub)
ah <- AnnotationHub()
xx <- ah[["AH99002"]]
```

2. For the second problem, which seems to fail at the unit test 'test_TEtranscripts()', could you please run the following lines and post the output of the last one:

```
library(atena)
bamfiles <- list.files(system.file("extdata", package="atena"),
                            pattern="*.bam", full.names=TRUE)
TE_annot <- readRDS(file = system.file("extdata", "Top28TEs.rds",
                                            package="atena"))
ttpar <- TEtranscriptsParam(bamfiles, teFeatures = TE_annot,
                               singleEnd = TRUE, ignoreStrand=TRUE,
                               aggregateby = c("repName"))
ttSE <- qtext(ttpar)
print(sort(assay(ttSE), decreasing=TRUE))
```

few weeks ago we actually had this problem with the BBS and in this case @beacalvo was following it up and she resolved it, so once we see your output, she may have a hint about what may be going on.

**Comment by martin-g** [comment]
About 1) there was a problem with the local cache of AnnotationHub (https://github.com/Bioconductor/AnnotationHub/issues/40). I will re-test now, but I am 99% certain it will solve this issue!

**Comment by martin-g** [comment]
Here is the new output of `CMD check`:
```
R CMD check atena_1.5.10.tar.gz 
* using log directory ‘/home/biocbuild/git/atena.Rcheck’
* using R Under development (unstable) (2023-03-12 r83975)
* using platform: aarch64-unknown-linux-gnu (64-bit)
* R was compiled by
    gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
    GNU Fortran (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
* running under: Ubuntu 22.04.2 LTS
* using session charset: UTF-8
* checking for file ‘atena/DESCRIPTION’ ... OK
* checking extension type ... Package
* this is package ‘atena’ version ‘1.5.10’
* package encoding: UTF-8
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking whether package ‘atena’ can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking ‘build’ directory ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking startup messages can be suppressed ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... OK
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking installed files from ‘inst/doc’ ... OK
* checking files in ‘vignettes’ ... OK
* checking examples ... OK
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ...
  Running ‘runTests.R’
 ERROR
Running the tests in ‘tests/runTests.R’ failed.
Last 13 lines of output:
  1 Test Suite : 
  atena RUnit Tests - 13 test functions, 0 errors, 1 failure
  FAILURE in test_TEtranscripts: Error in checkEqualsNumeric(head(sort(assay(ttSE), decreasing = TRUE)),  : 
    Mean relative difference: 0.007352941
  
  
  Test files with failing tests
  
     test_TEtranscripts.R 
       test_TEtranscripts 
  
  
  Error in BiocGenerics:::testPackage("atena") : 
    unit tests failed for package atena
  Execution halted
* checking for unstated dependencies in vignettes ... OK
* checking package vignettes in ‘inst/doc’ ... OK
* checking running R code from vignettes ...
  ‘atena.Rmd’ using ‘UTF-8’... OK
 NONE
* checking re-building of vignette outputs ... OK
* checking PDF version of manual ... OK
* DONE

Status: 1 ERROR
See
  ‘/home/biocbuild/git/atena.Rcheck/00check.log’
for details.
```

```
library(atena)
bamfiles <- list.files(system.file("extdata", package="atena"),
                            pattern="*.bam", full.names=TRUE)
TE_annot <- readRDS(file = system.file("extdata", "Top28TEs.rds",
                                            package="atena"))
ttpar <- TEtranscriptsParam(bamfiles, teFeatures = TE_annot,
                               singleEnd = TRUE, ignoreStrand=TRUE,
                               aggregateby = c("repName"))
ttSE <- qtext(ttpar)
print(sort(assay(ttSE), decreasing=TRUE))
```

fails with:
```
> ttSE <- qtext(ttpar)
Error in qtext(ttpar) : could not find function "qtext"
```

**Comment by martin-g** [comment]
It was a typo in your snippet:
```
> ttSE <- qtex(ttpar)
  |======================================================================| 100%

> print(sort(assay(ttSE), decreasing=TRUE))
[1] 149 123
```

**Comment by beacalvo** [comment]
Hi! As @rcastelo said, we came accross this error a few weeks ago but after some modifications it seemed fixed (did not prompt an error in R devel, nor in R release). The error was due to numeric instability during the EM algorithm. I have made a new modification in the code which should fix this problem. It is available on the development version of atena 1.5.11 which is now available in the atena github repo, or in 24/48 hours in the development branch of Bioconductor. Thank you!

**Comment by martin-g** [comment]
I can confirm that the latest commit fixed the problem:

```
biocbuild@kunpeng1 ~/git> R CMD build --keep-empty-dirs --no-resave-data atena
* checking for file ‘atena/DESCRIPTION’ ... OK
* preparing ‘atena’:
* checking DESCRIPTION meta-information ... OK
* installing the package to build vignettes
* creating vignettes ... OK
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building ‘atena_1.5.11.tar.gz’
```
```
biocbuild@kunpeng1 ~/git> R CMD check atena_1.5.11.tar.gz 
* using log directory ‘/home/biocbuild/git/atena.Rcheck’
* using R Under development (unstable) (2023-03-12 r83975)
* using platform: aarch64-unknown-linux-gnu (64-bit)
* R was compiled by
    gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
    GNU Fortran (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
* running under: Ubuntu 22.04.2 LTS
* using session charset: UTF-8
* checking for file ‘atena/DESCRIPTION’ ... OK
* checking extension type ... Package
* this is package ‘atena’ version ‘1.5.11’
* package encoding: UTF-8
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking whether package ‘atena’ can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking ‘build’ directory ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking startup messages can be suppressed ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... OK
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking installed files from ‘inst/doc’ ... OK
* checking files in ‘vignettes’ ... OK
* checking examples ... OK
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ...
  Running ‘runTests.R’
 OK
* checking for unstated dependencies in vignettes ... OK
* checking package vignettes in ‘inst/doc’ ... OK
* checking running R code from vignettes ...
  ‘atena.Rmd’ using ‘UTF-8’... OK
 NONE
* checking re-building of vignette outputs ... OK
* checking PDF version of manual ... OK
* DONE

Status: OK
```

Great work!
